Code/GraphMol/MolOps.cpp

```diff
@@ -1168,5 +1168,105 @@ std::vector<int> hapticBondEndpoints(const Bond *bond) {
   return oats;
 }
 }  // end of namespace details
-};  // end of namespace MolOps
-};  // end of namespace RDKit
+
+void expandAttachmentPoints(RWMol &mol, bool addAsQueries) {
+  for (auto atom : mol.atoms()) {
+    int value;
+    if (atom->getPropIfPresent(common_properties::molAttachPoint, value)) {
+      std::vector<int> tgtVals;
+      if (value == 1 || value == -1) {
+        tgtVals.push_back(1);
+      }
+      if (value == 2 || value == -1) {
+        tgtVals.push_back(2);
+      }
+      if (tgtVals.empty()) {
+        BOOST_LOG(rdWarningLog)
+            << "Invalid value for molAttachPoint: " << value << " on atom "
+            << atom->getIdx() << ". Not expanding this atttachment point."
+            << std::endl;
+        continue;
+      }
+      for (auto tval : tgtVals) {
+        atom->clearProp(common_properties::molAttachPoint);
+        Atom *newAtom = nullptr;
+        if (addAsQueries) {
+          newAtom = new QueryAtom(0);
+          newAtom->setQuery(RDKit::makeAtomNullQuery());
+        } else {
+          newAtom = new Atom(0);
+        }
+        newAtom->setProp(common_properties::_fromAttachPoint, tval);
+        bool updateLabel = false;
+        bool takeOwnership = true;
+        auto idx = mol.addAtom(newAtom, updateLabel, takeOwnership);
+        mol.addBond(atom->getIdx(), idx, Bond::SINGLE);
+      }
+    }
+  }
+}
+
+void collapseAttachmentPoints(RWMol &mol, bool markedOnly) {
+  bool removedAny = false;
+  std::vector<int> attachLabels(mol.getNumAtoms(), 0);
```

Comment:
Should we add a check on the labels we have seen on the mol? E.g.:
- We have seen 2 labels at most, and if so, these must be "1" and "2".
- If saw just 1 label, then it has to be either "1" or "-1".

Reply:
I thought about this, but it seems like we don't need to be that strict here?

Reply:
ok